File: src/test/java/swf/army/mil/blog/comment/CommentServiceTest.java
package swf.army.mil.blog.comment;

import jakarta.persistence.EntityNotFoundException;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class CommentServiceTest {

    @Mock
    CommentRepository commentRepository;

    @InjectMocks
    CommentService commentService;

    @Test
    void shouldSaveNewComment() {
        Comment testComment = new Comment(1L, null, null,
                "Author", "Content");

        when(commentRepository.save(testComment)).thenReturn(testComment);

        Comment actualComment = commentRepository.save(testComment);
        verify(commentRepository, times(1)).save(any(Comment.class));
        assertThat(testComment).isEqualTo(actualComment);
    }

    @Test
    void shouldFindCommentByID() {
        Long myID = 1L;

        Comment testComment = new Comment(myID, null, null,
                "Author", "Content");

        when(commentRepository.findById(myID)).thenReturn(Optional.of(testComment));

        Comment actualComment = commentService.findByID(myID);
        verify(commentRepository, times(1)).findById(any(Long.class));
        assertThat(testComment).isEqualTo(actualComment);
    }

    @Test
    void shouldNotFindIfDoesntExist() {
        Long missingID = 1L;

        when(commentRepository.findById(missingID)).thenReturn(Optional.empty());

        verify(commentRepository, never()).findById(missingID);

        EntityNotFoundException exception = assertThrows(EntityNotFoundException.class, () -> {
            commentService.findByID(missingID);
        });

        assertEquals("Comment not found", exception.getMessage());
    }

    @Test
    void shouldFindCommentsByPostID() {
        Long postID = 1L;

        List<Comment> testComments = List.of(new Comment(1L, null, null,
                "Author 1", "Content 1"),
                new Comment(2L, null, null,
                        "Author 2", "Content 2"));

        when(commentRepository.findByPostId(postID)).thenReturn(testComments);

        List<Comment> actualComments = commentService.findByPostID(postID);
        verify(commentRepository, times(1)).findByPostId(any(Long.class));
        assertThat(testComments).isEqualTo(actualComments);
    }

    @Test
    void shouldDeleteCommentByID() {
        Long myID = 1L;

        Comment testComment = new Comment(myID, null, null,
                "Author", "Content");

        when(commentRepository.findById(myID)).thenReturn(Optional.of(testComment));

        commentService.delete(myID);

        verify(commentRepository).delete(testComment);
    }

    @Test
    void shouldNotDeleteIfDoesntExist() {
        Long missingID = 1L;

        when(commentRepository.findById(missingID)).thenReturn(Optional.empty());

        verify(commentRepository, never()).delete(any());

        assertThrows(EntityNotFoundException.class, () -> commentService.delete(missingID));
    }
}

File: src/test/kotlin/com/personal/blog/comment/CommentServiceTest.kt
package com.personal.blog.post

import com.personal.blog.comment.Comment
import com.personal.blog.comment.CommentRepository
import com.personal.blog.comment.CommentService
import com.personal.blog.comment.CreateCommentRequest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.mockito.kotlin.*
import java.util.*

class CommentServiceTest {

    private val repo: CommentRepository = mock()
    private val postService: PostService = mock()
    private val service = CommentService(repo, postService)

    val req = CreateCommentRequest(1L, "Hello")
    val testComment = Comment(
        parentPost = Post(1L, "Test", "Post"),
        id = 1L,
        text = "Hello"
    )

    @Test
    fun `should save and return new comment`() {
        whenever(postService.retrieve(any()))
            .thenReturn(Post(1L, "Hello", "World"))
        whenever(repo.save(any())).thenReturn(testComment)

        val result = service.create(req)

        assertEquals(1L, result.id)
        assertEquals("Hello", result.text)

        verify(repo).save(any())
    }

    @Test
    fun `should return all comments on post`() {
        val posts = listOf(testComment)

        whenever(repo.findAllByParentPost_Id(any())).thenReturn(posts)

        val result = service.retrieveByPostID(1L)

        assertEquals(1, result.size)

        verify(repo).findAllByParentPost_Id(any())
    }

    @Test
    fun `should return comment by id`() {
        whenever(repo.findById(any())).thenReturn(Optional.of(testComment))

        val result = service.retrieveByID(1L)

        assertEquals("Hello", result.text)
        verify(repo).findById(any())
    }

    @Test
    fun `should throw if not found on retrieve`() {
        whenever(repo.findById(any())).thenReturn(Optional.empty())

        assertThrows(NoSuchElementException::class.java) {
            service.retrieveByID(99L)
        }

        verify(repo).findById(any())
    }

    @Test
    fun `should delete comment`() {
        whenever(repo.findById(any())).thenReturn(Optional.of(testComment))

        service.delete(1L)

        verify(repo).findById(any())
        verify(repo).delete(testComment)
    }

    @Test
    fun `should throw if not found on delete`() {
        whenever(repo.findById(any())).thenReturn(Optional.empty())

        assertThrows(NoSuchElementException::class.java) {
            service.delete(1L)
        }

        verify(repo).findById(any())
        verify(repo, never()).delete(any())
    }
}
